### PR TITLE
ovn-k8s-cni-overlay: Add Windows Server 2016 RTM support

### PR DESCRIPTION
**Pull request by alinbalutoiu**
This patch adds Windows Server 2016 RTM (build number 14393).

Kubernetes supports only one container per pod in this mode.
This Windows version does not support setting custom
IP/MAC address (will be supported by OVS on Windows in the near
future).

The CNI will update the port from OVN with the assigned IP/MAC address.

Signed-off-by: Alin Balutoiu <abalutoiu@cloudbasesolutions.com>

### CODE REVIEW

**Comment by alinbalutoiu** [comment]
This change depends on this PR from Kubernetes https://github.com/kubernetes/kubernetes/pull/61940
Waiting for the PR to get merged on Kubernetes before merging this PR.

File: go-controller/cmd/ovn-k8s-cni-overlay/app/helper_windows.go [hunk 1]
```diff
@@ -209,31 +235,210 @@ func PlatformSpecificCleanup(args *skel.CmdArgs) error {
...
+
+// Retrieves the Windows build number using wmic
+func getWindowsVersion() (int, error) {
+	args := []string{"os", "get", "BuildNumber", "/value"}
+	out, err := exec.Command("wmic", args...).Output()
+	if err != nil {
+		return 0, err
+	}
+	buildInfo := strings.Split(string(strings.Trim(string(out), "\" \n\r")), "=")
+	windowsVersion, _ := strconv.Atoi(buildInfo[1])
```
Comment (rajatchopra) [inline go-controller/cmd/ovn-k8s-cni-overlay/app/helper_windows.go]: We should not ignore the 'error' returned by strconv.Atoi, I think.

**Comment by rajatchopra** [comment]
Will this set of changes mean that we can run OVN +kubernetes on builds earlier than 1709?

File: go-controller/cmd/ovn-k8s-cni-overlay/app/helper_windows.go (hunk 1 above)
Comment (alinbalutoiu) [reply go-controller/cmd/ovn-k8s-cni-overlay/app/helper_windows.go]: I don't think the build number can be something else than an integer, but I will add the error handling anyway.

**Comment by alinbalutoiu** [comment]
@rajatchopra yes, that is what we want. I have prepared also an update to the kubernetes documentation to reflect all the changes (currently available here https://github.com/alinbalutoiu/kubernetes.github.io/blob/master/docs/getting-started-guides/windows/index.md#for-3-open-vswitch-ovs--open-virtual-network-ovn-with-overlay).

Those changes depend on this PR from kubernetes https://github.com/kubernetes/kubernetes/pull/61940

There will still be the limitation of one container per pod in builds earlier than 1709 unfortunately. We are still investigating on how to add support for multiple containers per pod in the earlier builds than 1709.

**Comment by alinbalutoiu** [comment]
Still waiting for the dependent PR to get merged in Kubernetes: https://github.com/kubernetes/kubernetes/pull/61940

**Comment by alinbalutoiu** [comment]
The patch in Kubernetes got merged.

This PR adds support for Windows Server 2016 build number 14393 in the CNI plugin and also adds the support inside the ansible playbooks to deploy it.

File: go-controller/cmd/ovn-k8s-cni-overlay/app/helper_windows.go [hunk 2]
```diff
@@ -209,31 +233,214 @@ func PlatformSpecificCleanup(args *skel.CmdArgs) error {
...
+			{
+				Version:   "4",
+				Interface: current.Int(1),
+				Address:   net.IPNet{IP: ipAddr, Mask: ipAddrNet.Mask},
+				Gateway:   net.ParseIP(gatewayIP),
+			},
+		},
+	}
+	// Since Windows Server 2016 doesn't support custom IP/Mac address, we have to update OVN with the given IP/Mac
+	stdout, stderr, err := util.RunOVNNbctl("--", "lsp-set-addresses", ifaceID, fmt.Sprintf("%s %s", macAddress, ipAddr.String()))
```
Comment (rajatchopra) [inline go-controller/cmd/ovn-k8s-cni-overlay/app/helper_windows.go]: will this conflict with ovnkube master's IPAM for the switch?

**Comment by alinbalutoiu** [comment]
Removed the ansible playbook for the moment as it will conflict with the other update to the playbooks. The ansible changes will be done in a separate PR.

**Comment by alinbalutoiu** [comment]
As discussed in one of the sig-windows meetings, older Windows versions support will be dropped out from Kubernetes. Closing this PR.

**Pull request closed without merging**